firewall.go

```diff
@@ -139,3 +667,67 @@ func FirewallRuleExistsByName(rules *ole.IDispatch, name string) (bool, error) {
 
 	return false, nil
 }
+
+// firewallRulesEnum takes fwPolicy object and returns all objects which needs freeing and enum itself,
+// which is used to enumerate rules. do not forget to:
+//   defer firewallRulesEnumRealease(ur, ep)
+func firewallRulesEnum(fwPolicy *ole.IDispatch) (*ole.VARIANT, *ole.VARIANT, *ole.IEnumVARIANT, error) {
+	unknownRules, err := oleutil.GetProperty(fwPolicy, "Rules")
+	if err != nil {
+		return nil, unknownRules, nil, fmt.Errorf("failed to get Rules: %s", err)
+	}
+	rules := unknownRules.ToIDispatch()
+
+	enumProperty, err := rules.GetProperty("_NewEnum")
+	if err != nil {
+		unknownRules.Clear()
+		return nil, unknownRules, nil, fmt.Errorf("failed to get enumeration property on Rules: %s", err)
+	}
+
+	enum, err := enumProperty.ToIUnknown().IEnumVARIANT(ole.IID_IEnumVariant)
+	if err != nil {
+		enumProperty.Clear()
+		unknownRules.Clear()
+		return nil, unknownRules, nil, fmt.Errorf("failed to cast enum to correct type: %s", err)
+	}
+	if enum == nil {
+		enumProperty.Clear()
+		unknownRules.Clear()
+		return nil, unknownRules, nil, fmt.Errorf("can't get IEnumVARIANT, enum is nil")
+	}
+	return unknownRules, enumProperty, enum, nil
+}
+
+// firewallRuleEnumRelease will free memory used by firewallRulesEnum.
+func firewallRulesEnumRealease(unknownRules, enumProperty *ole.VARIANT) {
+	enumProperty.Clear()
+	unknownRules.Clear()
+}
+
+// apiInit initialize common fw api.
+// then:
+// dispatch apiRelease(u, fwp)
+func apiInit() (*ole.IUnknown, *ole.IDispatch, error) {
+	ole.CoInitializeEx(0, ole.COINIT_APARTMENTTHREADED|ole.COINIT_SPEED_OVER_MEMORY)
+
+	unknown, err := oleutil.CreateObject("HNetCfg.FwPolicy2")
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to create FwPolicy Object: %s", err)
+	}
+
+	fwPolicy, err := unknown.QueryInterface(ole.IID_IDispatch)
+	if err != nil {
+		unknown.Release()
+		return nil, nil, fmt.Errorf("Failed to create FwPolicy Object (2): %s", err)
+	}
+
+	return unknown, fwPolicy, nil
+
+}
+
+// apiRelease cleans memory.
+func apiRelease(u *ole.IUnknown, fwp *ole.IDispatch) {
```

Comment:
same as above, probably best to add some "firewall" naming here since it is in the global namespace.

Reply:
done